colossalai/zero/low_level/low_level_optim.py

```diff
@@ -213,7 +217,11 @@ def _create_master_param_current_rank(self, param_list):
                     padding_param = param.data.view(-1)
                 splited_params = padding_param.split(padding_param.numel() // self._world_size)
 
-                splited_param_current_rank = splited_params[self._local_rank].detach().float().to(device)
+                # use fp32 when master_weights is True
+                if self._master_weights is True:
+                    splited_param_current_rank = splited_params[self._local_rank].detach().float().to(device)
+                else:
+                    splited_param_current_rank = splited_params[self._local_rank].detach().to(self._dtype).to(device)
```

Comment:
If no master weights, you should avoid data copy. This may lead to unneeded memory usage.